### ISSUE TITLE
Removing the physical optimizer `AddMergeExec` breaks the sorts

### ISSUE THREAD

**Issue by jorgecarleitao** [apache/datafusion#438]
AFAIK physical optimizers should be entirely optional and having them or not must result in the same semantics.

However, removing `AddMergeExec` breaks THC 1 atm (all sorts, actually) with

```
Error: Internal("SortExec requires a single input partition")
```

**Comment by Dandandan** [comment]
Yes, agreed 👍 I think hash aggregates might have the same problem.

Back when I added the physical optimizers I added this rule to keep it functioning the same way, as this was contained in the optimization pass.
I believe there were some comments back then too that we should change it at some time.

I think we have a couple of options:

* Add a `MergeExec` during planning time, wrap any input plan for a node that requires a single partition with `MergeExec` (to me it seems like the best option).
* Use `MergeExec` inside the `SortExec` - this is something done in the `HashJoinExec` and `CrossJoinExec` too. I am not sure if that's really elegant as it "hides" nodes from the plan tree - so maybe that should move to the planner too.

**Comment by alamb** [comment]
I think this ticket is no longer tracking anything actionable -- please reopen it (or a new ticket) if there is still work to be done here
